src/HealthChecks.EventStore/EventStoreHealthCheck.cs

```diff
@@ -4,89 +4,103 @@
 
 namespace HealthChecks.EventStore
 {
-    public class EventStoreHealthCheck : IHealthCheck
+    public class EventStoreHealthCheck : IHealthCheck, IDisposable
     {
         private const string CONNECTION_NAME = "AspNetCore HealthCheck Connection";
         private const int ELAPSED_DELAY_MILLISECONDS = 500;
         private const int RECONNECTION_LIMIT = 1;
 
-        private readonly string _eventStoreConnection;
-        private readonly string? _login;
-        private readonly string? _password;
+        private readonly IEventStoreConnection _eventStoreConnection;
 
+        private readonly TaskCompletionSource<HealthCheckResult> _tcs = new(TaskCreationOptions.RunContinuationsAsynchronously);
         public EventStoreHealthCheck(string eventStoreConnection, string? login, string? password)
+        {
+            Guard.ThrowIfNull(eventStoreConnection, throwOnEmptyString: true);
+            _eventStoreConnection ??= CreateEventStoreConnection(eventStoreConnection, login, password);
+        }
+
+        public EventStoreHealthCheck(IEventStoreConnection eventStoreConnection)
         {
             _eventStoreConnection = Guard.ThrowIfNull(eventStoreConnection);
-            _login = login;
-            _password = password;
         }
 
         public async Task<HealthCheckResult> CheckHealthAsync(HealthCheckContext context, CancellationToken cancellationToken = default)
         {
             try
             {
-                ConnectionSettingsBuilder connectionSettings;
-
-                if (string.IsNullOrEmpty(_login) || string.IsNullOrEmpty(_password))
-                {
-                    connectionSettings = ConnectionSettings.Create()
-                        .LimitReconnectionsTo(RECONNECTION_LIMIT)
-                        .SetReconnectionDelayTo(TimeSpan.FromMilliseconds(ELAPSED_DELAY_MILLISECONDS))
-                        .SetCompatibilityMode("auto");
-                }
-                else
-                {
-                    connectionSettings = ConnectionSettings.Create()
-                        .LimitReconnectionsTo(RECONNECTION_LIMIT)
-                        .SetReconnectionDelayTo(TimeSpan.FromMilliseconds(ELAPSED_DELAY_MILLISECONDS))
-                        .SetDefaultUserCredentials(new UserCredentials(_login, _password))
-                        .SetCompatibilityMode("auto");
-                }
-
-                using (var connection = EventStoreConnection.Create(
-                    _eventStoreConnection,
-                    connectionSettings,
-                    CONNECTION_NAME))
-                {
-                    var tcs = new TaskCompletionSource<HealthCheckResult>();
-
-                    //connected
-                    connection.Connected += (s, e) => tcs.TrySetResult(HealthCheckResult.Healthy());
+                //connected
+                _eventStoreConnection.Connected += (s, e) => _tcs.TrySetResult(HealthCheckResult.Healthy());

```

Comment:
You subscribe each time without unsubscribe. OOM here because of eternally growing list of delegates.

Reply:
Also such code is not thread safe.

Reply:
And that may be a reason why connection was created each time.

Reply:
Has `EventStoreConnection` some method to just call it? `Read()/Ping()/GetProperties()` ?

Reply:
Ah right.

> Has `EventStoreConnection` some method to just call it? `Read()/Ping()/GetProperties()` ?

Unfortunately not.

I can unsubscribe before returning, or perhaps rely on ConnectAsync alone.

Reply:
Alright, that should be fixed now, I updated the code to unsubscribe before returning.

Reply:
😕 no... it just does not work. Look - you try to subscribe/unsubscribe to events of shared connnection. Imagine 2-5-10 concurrent calls to your health check - all of them subscribe and work with the same `TaskCompletionSource`. Moreover - this `TaskCompletionSource` is not recreated. And moreover - you unsubscribe from events using each time **new** delegate. Unsubscription does not work.
```c#
 public class Button
    {
        public event EventHandler Click;
    }

    private EventHandler DoSomething(object context)
    {
        return (o, e) => { Console.WriteLine(context); };
    }
...
  var button = new Button();
        button.Click += DoSomething(1);
        button.Click -= DoSomething(1);  <----- button.Click is not null here
```

Reply:
Alright :/ damn. I think it's time to abandon this PR.

Thanks again for your help, I'm learning a lot from these PRs :D

Reply:
`EventStoreConnection` design contradicts with well-known pattern of long-lived health checks. All that events play not good in concurrent environment - too difficult to jump around synchronization constructions.